lib/srv/alpnproxy/proxy.go

```diff
@@ -530,3 +556,41 @@ func (p *Proxy) Close() error {
 	}
 	return nil
 }
+
+// MakeConnectionHandler creates a ConnectionHandler which provides a callback
+// to handle incoming connections by this ALPN proxy server.
+func (p *Proxy) MakeConnectionHandler(opts ...ConnectionHandlerOption) ConnectionHandler {
+	options := p.getHandleConnOptions(opts...)
+	return func(ctx context.Context, conn net.Conn) error {
+		return p.handleConn(ctx, conn, options)
+	}
+}
+
+// MakeDefaultTLSConfig clones provided TLS config and applies necessary
+// upgrades.
+func (p *Proxy) MakeDefaultTLSConfig(config *tls.Config) *tls.Config {
+	configClone := config.Clone()
+
+	// Add all supported protos.
+	configClone.NextProtos = apiutils.Deduplicate(append(configClone.NextProtos, common.ProtocolsToString(p.supportedProtocols)...))
+
+	// Do not ignore client certs if given. Some logic also requires
+	// information from the verified client cert.
+	configClone.ClientAuth = tls.VerifyClientCertIfGiven
+
+	// Set proper callback for fetching cluster CAs used for veriying clients.
+	configClone.GetConfigForClient = auth.WithClusterCAs(configClone, p.cfg.AccessPoint, p.cfg.ClusterName, p.cfg.Log)
```

Comment:
`WithClusterCAs` loads CAs based on SNI. In case of ALPN local proxy the SNI is always set to public proxy web address like `teleport.example.com` in the result WithClusterCAs will need to load all trusted CA for each incoming connection and in case of large numbers of trusted cluster can cause a pref issues. 

Wonder what is the real scenario for validation of UserCerts agains trusted cluster CA other than root CA ?

Reply:
You are totally right on this one. I will fall back to use the same CAs current web is using.